File: entregables/tp9.py
def prim(adj):
    Vt = ['0000']
    Xt = []
    for i in range(len(adj) - 1):
        min_e = None
        min_v = None
        for v in Vt:
            for vecino in adj[v]:
                if vecino[0] in Vt: continue
                if min_e == None or min_e[1] > vecino[1]:
                    min_e = vecino
                    min_v = vecino[0]
        
        Xt.append(min_e)
        Vt.append(min_v)
    return Vt, Xt

def dist(v, w):
    acc = 0
    for i in range(len(v)):
        c1, c2 = int(v[i]), int(w[i])
        diff = abs(c1 - c2)
        if diff <= 5:
            acc += diff
        else:
            acc += 10 - diff
    return acc

cases = int(input())
for k in range(cases):
    i = input()
    parsed = ['0000'] + i.split(' ')[1:]
    adj = {}

    for v in parsed:
        for w in parsed:
            if v == w: continue
            d = dist(v, w)
            if v not in adj:
                adj[v] = [(w, d)]
            else:
                adj[v].append((w, d))

    Vt, Xt = prim(adj)
    min_num = sum([e[1] for e in Xt])
    print(min_num)

File: entregables/tp9.cpp
#include <iostream>
#include <vector>
#include <list>
#include <string>
#include <utility> 
#include <algorithm>
#include <climits>
#include <queue>
#include <functional>
#define heap priority_queue
using namespace std;

int dist(string v, string w) {
    int acc = 0;
    for (int i = 0; i < v.size(); ++i) {
        int c1 = v[i] - '0', c2 = w[i] - '0';
        int diff = abs(c1 - c2);
        acc += min(diff, 10 - diff);
    }
    return acc;
}

int prim(const vector<list<pair<int, int>>>& adj, int start) {
    int n = adj.size();
    vector<int> minDist(n, INT_MAX);
    vector<bool> visited(n, false);
    // el heap tiene guardado el indice del nodo vecino en el array de nodos y el peso
    heap<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> min_heap;

    minDist[start] = 0;
    min_heap.push({0, start});

    int totalCost = 0;
    
    while (!min_heap.empty()) {
        int u = min_heap.top().second;
        min_heap.pop();

        if (visited[u]) continue;
        visited[u] = true;
        totalCost += minDist[u];

        for (const auto& [v, weight] : adj[u]) {
            if (!visited[v] && weight < minDist[v]) {
                minDist[v] = weight;
                min_heap.push({weight, v});
            }
        }
    }

    return totalCost;
}

int main() {
    int cases;
    cin >> cases;

    while (cases--) {
        int n;
        cin >> n;
        vector<string> nodes(n);
        for (int i = 0; i < n; ++i) {
            cin >> nodes[i];
        }

        int cantidad_minima_de_arranque = INT_MAX;
        for (int i = 0; i < n; ++i) {
            cantidad_minima_de_arranque = min(cantidad_minima_de_arranque, dist("0000", nodes[i]));
        }

        vector<list<pair<int, int>>> adj(n);
        
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                int d = dist(nodes[i], nodes[j]);
                adj[i].push_back({j, d});
                adj[j].push_back({i, d});
            }
        }

        int result = prim(adj, 0) + cantidad_minima_de_arranque;
        cout << result << endl;
    }

    return 0;
}
